mcpserver/tools.go

```diff
@@ -75,30 +82,52 @@ func RunStdioServer(ctx context.Context, dag *dagger.Client) error {
 	return stdioSrv.Listen(ctx, os.Stdin, os.Stdout)
 }
 
-var tools = []*Tool{}
+var tools = []MCPHandler{}
 
-func registerTool(tool ...*Tool) {
-	for _, t := range tool {
-		tools = append(tools, wrapTool(t))
-	}
+func registerTool[R Request, V Response](tool *Tool[R, V]) {
+	tools = append(tools, createHandler(tool))
 }
 
-func wrapTool(tool *Tool) *Tool {
-	return &Tool{
+func createHandler[R Request, V Response](tool *Tool[R, V]) MCPHandler {
+	return MCPHandler{
 		Definition: tool.Definition,
 		Handler: func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error) {
 			slog.Info("Tool called", "tool", tool.Definition.Name)
 			defer func() {
 				slog.Info("Tool finished", "tool", tool.Definition.Name)
 			}()
-			return tool.Handler(ctx, request)
+			// FIXME(aluzzardi): Backward compat, remove this
+			if tool.Handler != nil {
+				return tool.Handler(ctx, request)
+			}
+			parsed, err := tool.ParseRequest(request)
+			if err != nil {
+				return nil, err
+			}
+			response, err := tool.ActualHandler(ctx, parsed)
+			if err != nil {
+				return mcp.NewToolResultError(err.Error()), nil
+			}
+
+			json, err := json.Marshal(response.Data)
```

Comment:
🐛 **Bug (Critical):** Missing critical imports for 'json' and 'fmt' packages that are used in the json.Marshal and fmt.Errorf calls, causing compilation failure

```suggestion
import (
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"

	"dagger.io/dagger"
	"github.com/dagger/container-use/environment"
	"github.com/dagger/container-use/repository"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)
```